### ISSUE TITLE
LLM 学习路线：从模型原理到生产级 AI 应用

### ISSUE THREAD

**Issue by penn5012** [penn5012/penn.github.io#2]
# LLM 学习路线：从模型原理到生产级 AI 应用

> 面向已有 Node.js / TypeScript 开发经验、希望独立设计、开发、调试和上线 LLM 应用的开发者。
> GitHub Issue 不能创建真实文件夹，因此使用“目录 + 锚点详情”模拟文件夹结构。点击目录中的条目即可跳转到本 Issue 内对应知识点。

### 📁 00-基础认知
- [00. 学习目标与全局地图](#llm-overview)
- [01. AI、机器学习、深度学习与 LLM](#llm-definition)
- [02. LLM 能做什么，不能做什么](#llm-boundaries)

### 📁 01-模型原理
- [03. Token 与上下文窗口](#llm-token)
- [04. Embedding 与向量](#llm-embedding)
- [05. Transformer 与 Attention](#llm-transformer)
- [06. 自回归生成、Logits 与采样](#llm-generation)
- [07. 预训练、SFT 与对齐](#llm-training)

### 📁 02-API 与 Prompt
- [08. LLM API 请求结构](#llm-api)
- [09. Prompt 的组成与层级](#llm-prompt)
- [10. 多轮对话与上下文管理](#llm-context)
- [11. 结构化输出与结果校验](#llm-structured-output)
- [12. Streaming 流式输出](#llm-streaming)

### 📁 03-知识增强与行动能力
- [13. Embedding 搜索与 RAG](#llm-rag)
- [14. Function Calling / Tool Calling](#llm-tools)
- [15. Workflow、Agent 与 MCP](#llm-agent)
- [16. 多模态应用](#llm-multimodal)

### 📁 04-质量、安全与生产化
- [17. LLM 评测与回归测试](#llm-evaluation)
- [18. Prompt Injection 与安全](#llm-security)
- [19. 成本、延迟、限流与可靠性](#llm-reliability)
- [20. 可观测性与上线清单](#llm-production)

### 📁 05-AChat 实战
- [21. 当前 AChat LLM 调用链](#achat-current)
- [22. AChat 分阶段练习](#achat-practice)
- [23. 最终能力验收](#achat-acceptance)

---

<a id="llm-overview"></a>
## 00. 学习目标与全局地图
完成路线后，应能够解释 LLM 原理，使用 TypeScript/Node.js 调用模型，设计 Prompt、上下文、RAG、工具调用和 Agent，并处理评测、安全、成本、可观测性和上线问题。

~~~text
用户输入 → 前端 → 后端鉴权/校验 → Prompt/上下文/RAG/工具
→ LLM API → 输出校验 → 返回/持久化/观测
~~~

<a id="llm-definition"></a>
## 01. AI、机器学习、深度学习与 LLM
AI → ML → Deep Learning → Transformer → LLM。
AI 是总目标，机器学习从数据学习规律，深度学习使用神经网络，LLM 是主要处理自然语言、代码和其他序列数据的大型模型。LLM 不是传统数据库，也不是必然正确的搜索引擎，而是根据上下文生成 Token 的概率模型。

<a id="llm-boundaries"></a>
## 02. LLM 能做什么，不能做什么
适合文本生成、总结、分类、抽取、翻译、代码辅助、问答和工具编排。不应直接信任事实、数字、权限判断、SQL/Shell、JSON 格式或敏感信息。
> 模型输出必须当作不可信输入，服务端必须再次校验。

<a id="llm-token"></a>
## 03. Token 与上下文窗口
Token 是模型处理文本的基本单位，不严格等于汉字、单词或字符。文本会经过 Tokenizer 转为 Token ID，再参与模型计算。
Token 影响成本、延迟、上下文容量和历史消息压缩策略。上下文通常包括系统指令、用户消息、历史消息、工具描述、工具结果和 RAG 文档。
练习：统计长 Prompt 中重复内容，压缩后比较质量、Token 和延迟。

<a id="llm-embedding"></a>
## 04. Embedding 与向量
Embedding 把文本转换为向量，语义相近文本通常在向量空间中更接近。应用包括语义搜索、RAG、推荐、聚类、相似度检测和分类。
需要掌握余弦相似度、向量维度、文档分块、元数据过滤、向量数据库、更新和删除同步。

<a id="llm-transformer"></a>
## 05. Transformer 与 Attention
Transformer 的主链路是 Token → Embedding → 位置信息 → Self-Attention → MLP → Residual → LayerNorm → 多层 Block。
Attention 让模型决定当前 Token 应该关注上下文中的哪些 Token。Q/K/V 可以理解为查询、索引特征和实际内容。多头注意力可以同时学习语法、指代、时间、代码结构和长距离依赖。

<a id="llm-generation"></a>
## 06. 自回归生成、Logits 与采样
LLM 通常逐个预测下一个 Token。Logits 经过 Softmax 形成概率。需要理解 temperature、top-p、最大输出 Token、推理强度和输出详细程度。一次只改一个参数，并使用固定评测集比较。

<a id="llm-training"></a>
## 07. 预训练、SFT 与对齐
预训练从大量文本中学习语言和世界模式；SFT 使用人工问答示例学习任务、格式和语气；对齐让模型学习指令遵循、安全、隐私保护和拒答策略。
要区分 Prompt（运行时控制）、RAG（运行时提供知识）、Fine-tuning（改变模型参数）和基础模型训练。

<a id="llm-api"></a>
## 08. LLM API 请求结构
Node.js/TypeScript 最小调用：
~~~ts
import OpenAI from 'openai'
const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY })
const response = await client.responses.create({
  model: process.env.OPENAI_MODEL ?? 'gpt-5.4-nano',
  input: [{ role: 'user', content: '请解释 Node.js 事件循环' }],
})
console.log(response.output_text)
~~~
生产请求还要处理模型、Prompt、历史、超时、重试、取消、流式、结构化输出、工具、日志、Token 和成本。API Key 只能在服务端保存。

<a id="llm-prompt"></a>
## 09. Prompt 的组成与层级
高质量 Prompt 通常包含：角色 + 任务 + 上下文 + 约束 + 输出格式 + 验收标准。
需要掌握 system/developer/user 层级、zero-shot、few-shot、模板、动态变量、防 Prompt Injection、版本管理和评测。指令要明确、可验证，避免互相矛盾和重复堆叠。

<a id="llm-context"></a>
## 10. 多轮对话与上下文管理
LLM 默认没有长期记忆。应用需要读取历史、过滤无关内容、截取最近轮次、摘要旧消息，再发给模型。
必须处理上下文超限、消息顺序、user/assistant 配对、隐私删除、多用户隔离和并发写入。

<a id="llm-structured-output"></a>
## 11. 结构化输出与结果校验
不要只写“请返回 JSON”，应定义 Schema，并校验类型、枚举、必填字段和额外字段。需要处理解析失败、Schema 版本变化和重试。

<a id="llm-streaming"></a>
## 12. Streaming 流式输出
非流式是等待完整答案后一次返回；流式是逐片段返回。需要学习 SSE、WebSocket、AbortController、取消、断线重连、重复片段、部分消息保存、代理缓冲和连接超时。AChat 当前是非流式第一阶段。

<a id="llm-rag"></a>
## 13. Embedding 搜索与 RAG
离线：文档 → 清洗 → 分块 → Embedding → 向量数据库。在线：问题 → Embedding → 检索 → 权限过滤 → 重排序 → Prompt → LLM。
重点是 Chunk、召回数量、相似度阈值、元数据、来源引用、文档更新、权限、多租户和检索评测。

<a id="llm-tools"></a>
## 14. Function Calling / Tool Calling
模型提出工具调用，后端校验参数并执行函数，再把工具结果返回模型。必须具备参数 Schema、权限检查、超时、幂等、审计、用户确认和错误处理。禁止直接执行模型生成的 SQL、Shell 或删除命令。

<a id="llm-agent"></a>
## 15. Workflow、Agent 与 MCP
Chat 是输入 → 模型 → 输出；Workflow 的流程由代码固定；Agent 由模型动态决定下一步；MCP 通过标准协议暴露 Tools、Resources 和 Prompts。
流程固定优先 Workflow；只有步骤不固定时才引入 Agent。

<a id="llm-multimodal"></a>
## 16. 多模态应用
学习图片、音频、视频、文件、OCR 和图表输入，以及上传、MIME、大小、分辨率、Token、隐私、安全和异步任务。

<a id="llm-evaluation"></a>
## 17. LLM 评测与回归测试
建立固定测试集，评估正确率、任务完成率、格式符合率、引用准确率、工具调用成功率、幻觉率、延迟、Token、成本和用户满意度。修改模型、Prompt、工具或 RAG 后重新跑回归。

<a id="llm-security"></a>
## 18. Prompt Injection 与安全
防护：不把用户输入拼进系统规则；工具权限由后端决定；模型只能提出动作；敏感操作需要确认；服务端验证参数；日志脱敏；多租户隔离；API Key 不进前端和 Git。

<a id="llm-reliability"></a>
## 19. 成本、延迟、限流与可靠性
成本由输入 Token、输出 Token、模型价格、请求数、工具调用和检索文档数量决定。需要处理 400、401/403、429、500、网络中断、超时、空响应和非法输出。重试应限制次数、指数退避并避免副作用重复执行。

<a id="llm-production"></a>
## 20. 可观测性与上线清单
记录 requestId、provider、model、latency、Token、status、errorCode、toolCalls 和用户反馈；不要记录 API Key、密码和未脱敏隐私。
上线前检查 Key、超时、限流、重试、降级、错误映射、成本上限、健康检查、优雅停机、日志、指标、回归测试和数据删除方案。

<a id="achat-current"></a>
## 21. 当前 AChat LLM 调用链
frontend/src/App.tsx → frontend/src/api/client.ts → POST /api/conversations/:id/messages → conversation.route.ts → conversation.service.ts → model-provider-factory.ts → openai.provider.ts / deepseek.provider.ts。
关键思想：业务层依赖统一 ModelProvider 接口，而不是直接依赖某一个供应商。相关文件：
- [OpenAI Provider](https://github.com/penn5012/penn.github.io/blob/main/backend/src/providers/openai.provider.ts)
- [DeepSeek Provider](https://github.com/penn5012/penn.github.io/blob/main/backend/src/providers/deepseek.provider.ts)
- [Provider 工厂](https://github.com/penn5012/penn.github.io/blob/main/backend/src/providers/model-provider-factory.ts)
- [会话服务](https://github.com/penn5012/penn.github.io/blob/main/backend/src/modules/conversations/conversation.service.ts)
- [API 契约](https://github.com/penn5012/penn.github.io/blob/main/docs/api/openapi.yaml)

<a id="achat-practice"></a>
## 22. AChat 分阶段练习
1. 解释 API Key 安全和 400/502/503 的区别。
2. 增加教学型系统 Prompt。
3. 实现最近 10 轮上下文、旧消息摘要和 Token 预算。
4. 增加带 Schema 的学习计划生成接口。
5. 将 Node.js 文档导入 RAG，并返回来源引用。
6. 增加查询会话历史、搜索项目文档和健康检查工具。
7. 建立至少 20 个问题，比较不同 Provider、Prompt 和上下文策略。

<a id="achat-acceptance"></a>
## 23. 最终能力验收
- [ ] 能解释 Token、Embedding、Attention 和上下文窗口。
- [ ] 能独立用 TypeScript 调用 LLM。
- [ ] 能设计 Prompt、输出 Schema 和多轮上下文。
- [ ] 能实现 Streaming、RAG 和 Tool Calling。
- [ ] 能区分 Workflow、Agent 和 MCP。
- [ ] 能设计 Prompt Injection 防护。
- [ ] 能编写回归评测并监控 Token、延迟、错误和成本。
- [ ] 能安全管理 API Key，并处理超时、限流、重试和降级。
- [ ] 能解释 AChat 每一层代码的职责。

## 推荐学习顺序
模型原理 → API 和 Prompt → 上下文和结构化输出 → Streaming → Embedding 和 RAG → Tool Calling → Workflow / Agent / MCP → 评测、安全和生产化

## 参考资料
- [OpenAI API Model Guidance](https://developers.openai.com/api/docs/guides/latest-model)
- [OpenAI Models](https://developers.openai.com/api/docs/models)
- [OpenAI Embeddings](https://developers.openai.com/api/docs/models/text-embedding-3-large)
- [AChat Day 001](https://github.com/penn5012/penn.github.io/issues/1)
